dhis-2/dhis-services/dhis-service-tracker/src/main/java/org/hisp/dhis/tracker/validation/hooks/AbstractTrackerValidationHook.java

```diff
@@ -0,0 +1,393 @@
+
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+import java.util.stream.Collectors;
+
+import static org.hisp.dhis.tracker.report.ValidationErrorReporter.newReport;
+
+/**
+ * @author Morten Svanæs <msvanaes@dhis2.org>
+ */
+public abstract class AbstractTrackerValidationHook
+    implements TrackerValidationHook
+
+{
+    @Autowired
+    protected ProgramStageInstanceService programStageInstanceService;
+
+    @Autowired
+    protected IdentifiableObjectManager manager;
+
+    @Autowired
+    protected TrackedEntityInstanceService trackedEntityInstanceService;
+
+    @Autowired
+    protected EventService eventService;
+
+    @Autowired
+    protected FileResourceService fileResourceService;
+
+    @Autowired
+    protected TrackedEntityAttributeValueService trackedEntityAttributeValueService;
+
+    @Autowired
+    protected TrackedEntityAttributeService teAttrService;
+
+    @Autowired
+    protected ReservedValueService reservedValueService;
+
+    @Autowired
+    protected TrackedEntityProgramOwnerService trackedEntityProgramOwnerService;
+
+    @Autowired
+    protected TrackerOwnershipManager trackerOwnershipManager;
+
+    @Autowired
+    protected TrackedEntityCommentService commentService;
+
+    @Autowired
+    protected CurrentUserService currentUserService;
+
+    @Autowired
+    protected I18nManager i18nManager;
+
+    @Autowired
+    protected TrackerAccessManager trackerAccessManager;
+
+    @Autowired
+    protected TrackedEntityInstanceStore trackedEntityInstanceStore;
+
+    @Autowired
+    protected OrganisationUnitService organisationUnitService;
+
+    @Autowired
+    protected TrackedEntityTypeService defaultTrackedEntityTypeService;
+
+    @Autowired
+    public AclService aclService;
+
+    @Autowired
+    protected ProgramInstanceService programInstanceService;
+
+    @Autowired
+    protected RelationshipService relationshipService;
+
+    protected boolean validateAttributes( ValidationErrorReporter errorReporter, TrackerBundle bundle,
+        TrackedEntity te, TrackedEntityInstance tei, OrganisationUnit orgUnit )
+    {
+        ValidationErrorReporter reporter = new ValidationErrorReporter( bundle, errorReporter.getMainKlass() );
+        // Set main id the same for all errors in the report.
+        reporter.setMainId( te );
+        reporter.setLineNumber( errorReporter.getLineNumber() );
+
+        // For looking up existing tei attr. ie. if it is an update. Could/should this be done in the preheater instead?
+        Map<String, TrackedEntityAttributeValue> valueMap = getTeiAttributeValueMap(
+            trackedEntityAttributeValueService.getTrackedEntityAttributeValues( tei ) );
+
+        for ( Attribute attr : te.getAttributes() )
+        {
+            if ( StringUtils.isEmpty( attr.getValue() ) )
+            {
+                continue;
+            }
+            // no/empty value? should this qualify as an error?
```

Comment:
To be discussed, but sometimes this mean "clear the value", i.e. delete completely from the database

Reply:
Ok, then I will just do a "continue;" on empty values here.